modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/KafkaToIgniteCdcStreamerApplier.java

```diff
@@ -0,0 +1,253 @@
+ * @see CacheEntryVersion
+ */
+class KafkaToIgniteCdcStreamerApplier extends CdcEventsApplier implements Runnable, AutoCloseable {
+    /** */
+    public static final int DFLT_REQ_TIMEOUT = 3;
+
+    /** Ignite instance. */
+    private final IgniteEx ign;
+
+    /** Log. */
+    private final IgniteLogger log;
+
+    /** Closed flag. Shared between all appliers. */
+    private final AtomicBoolean closed;
+
+    /** Kafka properties. */
+    private final Properties kafkaProps;
+
+    /** Topic to read. */
+    private final String topic;
+
+    /** Lower kafka partition (inclusive). */
+    private final int kafkaPartFrom;
+
+    /** Higher kafka partition (exclusive). */
+    private final int kafkaPartTo;
+
+    /** Caches ids to read. */
+    private final Set<Integer> caches;
+
+    /** Consumers. */
+    private final List<KafkaConsumer<Integer, byte[]>> cnsmrs = new ArrayList<>();
+
+    /** */
+    private final AtomicLong rcvdEvts = new AtomicLong();
+
+    /**
+     * @param ign Ignite instance.
+     * @param log Logger.
+     * @param kafkaProps Kafka properties.
+     * @param topic Topic name.
+     * @param kafkaPartFrom Read from partition.
+     * @param kafkaPartTo Read to partition.
+     * @param caches Cache ids.
+     * @param maxBatchSize Maximum batch size.
+     * @param closed Closed flag.
+     */
+    public KafkaToIgniteCdcStreamerApplier(
+        IgniteEx ign,
+        IgniteLogger log,
+        Properties kafkaProps,
+        String topic,
+        int kafkaPartFrom,
+        int kafkaPartTo,
+        Set<Integer> caches,
+        int maxBatchSize,
+        AtomicBoolean closed
+    ) {
+        super(maxBatchSize);
+
+        this.ign = ign;
+        this.kafkaProps = kafkaProps;
+        this.topic = topic;
+        this.kafkaPartFrom = kafkaPartFrom;
+        this.kafkaPartTo = kafkaPartTo;
+        this.caches = caches;
+        this.closed = closed;
+        this.log = log.getLogger(KafkaToIgniteCdcStreamerApplier.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void run() {
+        U.setCurrentIgniteName(ign.name());
+
+        try {
+            for (int kafkaPart = kafkaPartFrom; kafkaPart < kafkaPartTo; kafkaPart++) {
+                KafkaConsumer<Integer, byte[]> cnsmr = new KafkaConsumer<>(kafkaProps);
+
+                cnsmr.assign(Collections.singleton(new TopicPartition(topic, kafkaPart)));
+
+                cnsmrs.add(cnsmr);
+            }
+
+            Iterator<KafkaConsumer<Integer, byte[]>> cnsmrIter = Collections.emptyIterator();
+
+            while (!closed.get()) {
+                if (!cnsmrIter.hasNext())
+                    cnsmrIter = cnsmrs.iterator();
+
+                poll(cnsmrIter.next());
+            }
+        }
+        catch (WakeupException e) {
+            if (!closed.get())
+                log.error("Applier wakeup error!", e);
+        }
+        catch (Throwable e) {
+            log.error("Applier error!", e);
+
+            closed.set(true);
```

Comment:
all other appliers should be waked up as well.

Reply:
I don't think so. Why?
All other appliers will try to perform last poll and will finish after.

Reply:
You're performing wake-up at another place when setting closed to true, why not here?